### DISCUSSION TITLE
Deferred control transfer completion

### DISCUSSION THREAD

**Discussion by smunaut** [hathach/tinyusb#1466] [Q&A]
So for a given application, I'd like to defer the decisions of OK/Error and not provide them directly as returns to the callbacks.

That is, once I get the SETUP stage, I can't immediately decide if it should stall or start a transfer. And for OUT transactions, when I get the DATA in the xfer completion callback, I can't immediately decide if the status stage should be OK or not.

How can I do that ?

(Basically I need to query another chip and only when I get the response from that other chip can I return the results ...)

**Comment by hathach** [comment]
can I ask which control transfer (specific request) you are working with, and how long you need to wait. Currently it is only true/false, but adding deferred(try-again) code is possible if it is common enough !!

**Reply by smunaut** [reply]
So in this context this is for the MCH2022 badge. Basically we're using the RP2040 as a "external usb processor" but there is an ESP32 which is the main application processor and so we'd like to basically forward all control request to a given interface (for WebUSB) to the ESP32 and then the ESP32 will take the decision of the result and do whatever. This way users can add custom control requests in their apps without altering the RP2040 firmware (which isn't meant to be user modifyable).

As to how long ... a few milisecond for the setup stage. Possibly a bit more if there are actual data to process / echange. I know USB has timeouts for this in the 50 ms range so obviously we need to fit in that.

**Reply by hathach** [reply]
would it be OK if you just forward data to ESP32 and get the response before returning to the callback. It is a bit blocking, but easiest way than upgrading/changing the current callback/driver. 
Meanwhile usb controller still NAKs next packet from host until the callback return. 

For timeout (50 ms or so as you said), if it is kernel driver/app timeout (even if we responded with NAK the whole time), there is no way around that, since even we enhance tinyusb, it just simply improve the mcu blocking on this specific usb events.

**Reply by smunaut** [reply]
Blocking isn't really an option. The RP2040 does other stuff and can't block in the usb stack.
Actually we wouldn't even see the response because the process that talks to the ESP32 is one of the other tasks that run and if we block in tinyusb it wouldn't get run.

**Reply by hathach** [reply]
sorry late response, in that case I think we need to change the return type of the control callback from bool to int. To stay backward-compatible, maybe having
- (-1) as try-again
- 0 or false as not supported 
- 1 or true as OK

**Reply by sigsegdammit** [reply]
Has there been any work on this?

I have a similar application where a control request kicks off an i2c transfer and I do not want to block.

Would a return code of -1 cause the control callback to be called again? I was thinking a -1 does nothing ( neither status nor stall ). Later my code would then be responsible to call either tud_control_xfer(), or a new function tud_control_stall().
